### PR TITLE
PT-3558: Fix scripture editor scrolling by BCV

### PR DESCRIPTION
**Pull request by lyonsil**
# Code Review Summary

**Branch**: fix/scroll-group-scroll-into-view

**Base**: origin/main

**Date**: 2026-07-10

**Review model**: Claude Fable 5

**Files changed**: 11 (plus README.md updated during the review)

## Overview

Fixes scrolling by BCV (book/chapter/verse) in the scripture editor, completing the fix for [PT-3558](https://paratextstudio.atlassian.net/browse/PT-3558). The root cause: the scroll code assumed `.editor-container` is the scrolling element, but unconstrained wrapper divs leave it auto-height, so it grows to its content height and scrolling it is a silent no-op — the web view's outer `tw:overflow-auto` wrapper is what actually scrolls. The fix discovers the real scroll container at scroll time (`findScrollContainer`) instead of assuming a selector, and switches the offset math from an `offsetParent` walk to `getBoundingClientRect` differences (an unpositioned scroll container may not appear in the offsetParent chain). The author chose runtime discovery deliberately as a durable fix: editor restyling has broken automatic scrolling twice this year, and discovery always lands on whatever container is actually vertically scrollable.

The branch also adds regression protection at two levels: a 21-test jsdom unit suite for the scroll utilities (including an explicit regression test mirroring the shipped bug) and a new isolated e2e spec (`scroll-group-sync.spec.ts`). Supporting e2e infrastructure: a `PLATFORM_BIBLE_PROJECT_ROOT_FOLDER` env override in `LocalParatextProjects.cs` so e2e runs use an isolated project root instead of the developer's real projects, fixture plumbing (`electronLaunchOptions` option fixture), and a subset runner (`e2e-tests/run-isolated.mjs`) behind `npm run test:e2e:isolated`.

## Round 2 — Follow-up fixes (commit f9bfb78c8bf)

A second, max-effort review pass surfaced a few defects in the tooling/e2e changes plus some hardening opportunities in the scroll utilities. These were fixed on top of the original branch. (The `PLATFORM_BIBLE_PROJECT_ROOT_FOLDER` production-gating discussed under Minor findings was deliberately left as-is per the author.)

- **`run-isolated.mjs`** — the earlier change dropped Playwright's positional path-filter passthrough, which broke the documented `npm run test:e2e:isolated -- <path>` invocation in `comment-test-helpers.ts` (it errored `Unknown subset`). Path passthrough is restored; a bare no-arg invocation now **exits non-zero** (was `0`) so it cannot report a false green; the duplicate arg-dispatch branches were merged.
- **`scrollToVerse` (`editor-dom.util.ts`)** — scroll-container discovery (a `getComputedStyle` + reflow ancestor walk) is now deferred inside the scroll guard, so `model-text-panel`'s per-frame rAF retry loop does no layout work on frames where the target verse marker has not painted yet.
- **`helpers.ts`** — an ambient `PLATFORM_BIBLE_PROJECT_ROOT_FOLDER` from the shell/CI is now stripped from the inherited env, so it can no longer leak into non-isolated e2e suites; only the `isolatedProjectRoot` opt-in sets it.
- **`scrollToAnnotation`** — the annotation id is `CSS.escape`d before interpolation into the class selector (matching `scripture-pane.component.tsx`), preventing a `SyntaxError` on ids containing CSS metacharacters.
- **`getTopWithinScrollContainer`** — subtracts the container's `clientTop` (top border) and takes the element rect as a parameter so `scrollToAnnotation` reads `getBoundingClientRect()` once.
- **`LocalParatextProjects.cs`** — the project-root override predicate is evaluated once (if/else) instead of twice.
- **Tests** — three new unit tests (**21** total, up from 18) cover the container-top offset, the negative-target clamp, and id escaping (with a jsdom `CSS.escape` polyfill matching `overlay-coordinates.test.ts`).

Round-2 verification: the 21-test `editor-dom.util` unit suite passes; extension + e2e ESLint clean (prettier-formatted); `dotnet build` succeeds with 0 errors and the C# file is csharpier-clean; pre-commit hooks (gitleaks, lint-staged) ran on commit.

## API Changes

None. No public API surface files changed:

- `lib/papi-dts/papi.d.ts` — untouched
- `lib/platform-bible-react/`, `lib/platform-bible-utils/` — untouched
- Extension type declaration files (`extensions/src/*/src/types/*.d.ts`) — untouched

The only new export is `findScrollContainer` in `extensions/src/platform-scripture-editor/src/editor-dom.util.ts`, an extension-internal module (not part of any declared API surface; consumed only within the extension and its tests).

### Critical — Must address before merge

None.

### Important — Should address before merge

- [x] **`package.json:101` / `e2e-tests/run-isolated.mjs`: bare `npm run test:e2e:isolated` changed behavior — it now prints the subset list and exits 0 without running any tests, while `README.md` still documented the old run-everything behavior; a scripted caller following the README would get a green exit code having run nothing.** _(fixed during review: README bullet updated to document the new usage — `all` runs every test, `<subset>` runs one feature, bare invocation lists subsets and runs nothing. Author confirmed the bare-invocation behavior change itself is intentional. CI is unaffected; workflows only run `test:e2e:smoke`. **Round 2:** the bare invocation now exits **non-zero** (not `0`) so it can't report a false green, and Playwright path-filter passthrough was restored — see the Round 2 section.)_

Author response: The behavior change was intentional; the author asked for the README to be updated, which was done in-review.

### Minor — Consider

- [x] **`c-sharp/Projects/LocalParatextProjects.cs:65-78`: the new `PLATFORM_BIBLE_PROJECT_ROOT_FOLDER` env override is active in production builds (not gated to dev/test), and was only documented in a code comment — a stray environment value silently redirects the app away from the real `Paratext 9 Projects` folder.** _(fixed during review: added a `Console.WriteLine` announcing the override and its path whenever the env var is set, so it shows up in the log. Author is explicitly comfortable shipping the override ungated in production builds.)_
- [x] **`editor-dom.util.ts:78-90`: `findScrollContainer` duplicated the ancestor walk in `paragraph-marker-tooltip-overlay.component.tsx` (same computed-style overflow loop, differing only in the `scrollHeight > clientHeight` check), with cross-referencing comments explaining the divergence.** _(fixed during review: `findScrollContainer` gained an optional `{ requireOverflow?: boolean }` parameter (default `true`); the tooltip overlay now imports it with `requireOverflow: false` instead of its hand-rolled walk, which also removed two `no-type-assertion` eslint-disables. Two new unit tests cover the style-only mode; all 21 tests pass (18 at review time, +3 in round 2).)_
- [x] **`e2e-tests/tests/isolated/scroll-groups/scroll-group-sync.spec.ts:35-39`: comment referenced a gitignored task brief ("DEVIATION from task brief... Task 3"), meaningless to future readers of the repo.** _(fixed during review: comment trimmed to the substantive rationale — Playwright's base `test` already registers a worker-scoped `launchOptions` fixture, hence the `electronLaunchOptions` name.)_
- [x] **`editor-dom.util.ts`: `scrollToVerse` and `scrollToAnnotation` each repeated the same rect-difference computation (`scrollTop + elementRect.top - containerRect.top`) with a "see scrollToVerse" cross-reference.** _(fixed during review: extracted a private `getTopWithinScrollContainer` helper that both functions share; the offsetParent-vs-rect rationale comment now lives in one place.)_

Author response: Author requested fixes for all four minor findings; all were implemented and verified during the review.

#### Shared Regions Modified

None. (Verified by grep: no changed file contains a `#region shared with` marker.)

#### Extension Config Changes

- [n/a] All extension changes are source files under `extensions/src/platform-scripture-editor/src/` — extension-specific, no config/build files touched, no propagation needed.
- [n/a] `package.json` (repo root), `e2e-tests/**` — not extension template files.

## Positive Observations

- The fix addresses the actual root cause and is applied consistently to both `scrollToVerse` and `scrollToAnnotation`, not just the BCV path named in the purpose; the chapter-start edge case (`verseNum <= 1`, no verse marker) preserves prior effective behavior (scroll to top).
- Two-level regression protection: a 21-test jsdom unit suite (including an explicit `REGRESSION:` test mirroring the shipped bug and an already-scrolled-container case) plus an e2e guard with a falsifiability precondition (`expect(lastVerse).not.toBeInViewport()` before the navigation under test).
- The isolated-project-root design keeps e2e runs off developers' real Paratext projects, reuses the existing `userDataDir` teardown for cleanup, and the C# env-var read follows the existing `DEV_NOISY` precedent rather than inventing a new configuration pattern.
- The e2e spec reuses existing fixture helpers (`sendPapiRequestOnce`, `waitForPapiMethodRegistered`, `waitForAtLeastOneProjectMetadata`) instead of reimplementing WebSocket plumbing, and its hardcoded sample-project GUID matches `c-sharp/assets/WEB/Settings.xml`.
- `run-isolated.mjs` is cross-platform (`process.execPath` + `require.resolve('@playwright/test/cli')`, no hardcoded paths) and exits non-zero on unknown subsets or a signal-killed child.
- Comments consistently explain _why_: the `_editor-overrides.scss` note documenting that its scroll styles are currently inert, the TSDoc on discovery-not-assumption, and the fixture-collision rationale at the definition site.
- UX-wise the change restores the documented BCV-navigation guideline ("Tabs... automatically follow the selected new scripture reference"); smooth scrolling and the `VERSE_NUMBER_SCROLL_OFFSET` context margin are preserved, and the rect math is vertical-only, so no RTL concerns.
- No user-visible strings added anywhere in the diff, so no localization obligations arise.

## Interview Notes

- **Stated purpose**: Fix scrolling by BCV; completes the fix for PT-3558.
- **Design rationale (runtime scroll-container discovery)**: The author explained this clearly — editor restyling silently broke automatic scrolling twice this year because the scroll code chased a specific selector. Discovering the scroll container at scroll time is a durable fix: it always lands on whatever container is actually vertically scrollable for the verse, so future layout changes resolve correctly either way.
- **Offset math (`offsetParent` walk → rect differences)**: The author deferred to the in-code comments rather than restating the reasoning. The comments do document it (an unpositioned scroll container may not appear in the offsetParent chain, so rect differences are used instead); the reviewer may want a quick walkthrough of that math in the meeting to confirm shared understanding.
- **`test:e2e:isolated` bare invocation**: Intentional design — it lists available subsets and runs nothing. The README was updated in-review to match.
- **Production env override**: The author is comfortable with `PLATFORM_BIBLE_PROJECT_ROOT_FOLDER` being active in production builds and chose a startup `Console.WriteLine` (visible in the log) over gating it to dev/test.
- All four minor findings were fixed at the author's request during the review. Nothing else was flagged in wrap-up.

## In-Review Quality Check

- `npm run typecheck` — passed cleanly.
- `npm run lint` — passed with 0 errors. Two prettier warnings in `editor-dom.util.ts` (from the in-review refactor) were resolved by the scoped format step; two pre-existing `no-console` warnings in `use-effective-resource-reference-list.hook.ts` are untouched by this branch and were left alone.
- Scoped prettier format — one TSDoc line-reflow in `editor-dom.util.ts`; no other branch files needed changes.
- `npm test` — all 783 TypeScript tests pass (round 2 later adds 3 unit tests → 786; see the Round 2 section). Two of three full-suite runs exited 1 due to a **pre-existing** load-dependent teardown flake (unhandled `document is not defined` rejection in `overlay.service-host.ts:150` surfacing from `overlay.service-host.contribution.test.ts`, code last touched in PRs #2173/#2106); the test file passes in isolation and a clean full-suite run completed. Unrelated to this branch; reported, not fixed.
- `dotnet test` (c-sharp-tests) — 1393 passed, 0 failed, 6 skipped. Note: the new `Console.WriteLine` is visible in C# test output when the env override is set — harmless, but a reviewer may notice the extra line.

## Suggested Review Focus

Prioritized areas for the author-reviewer meeting:

- [ ] The core design: `findScrollContainer` runtime discovery (including the new `requireOverflow` option now shared with `ParagraphMarkerTooltipOverlay`) — is the team comfortable standardizing on discovery over selector assumptions for editor scrolling?
- [ ] The rect-based offset math — the author deferred to the in-code comments for the rationale; worth a two-minute walkthrough to confirm shared understanding.
- [ ] `PLATFORM_BIBLE_PROJECT_ROOT_FOLDER` active in production builds — author's deliberate choice (with startup logging); confirm the team accepts the tradeoff versus gating it. (Round 2 stops an ambient value from leaking into non-isolated e2e suites, but the production gating itself is unchanged.)
- [ ] `npm run test:e2e:isolated` bare invocation lists subsets and — as of round 2 — exits non-zero (was `0`), with Playwright path-filter passthrough restored; the false-green concern is resolved. Confirm no scripted caller depended on the old run-everything behavior.

AI-assisted — [session](https://claude.ai/code/session_58813d30-de77-4c40-b920-2b73b2524901)

[PT-3558]: https://paratextstudio.atlassian.net/browse/PT-3558?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

- - -
This change is [Reviewable](https://reviewable.io/reviews/paranext/paranext-core/2533)

### CODE REVIEW

**Review by tjcouch-sil — APPROVED** [review]
LGTM thanks!

File: e2e-tests/tests/isolated/scroll-groups/scroll-group-sync.spec.ts [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+  // this exact command to be registered so the request cannot fail with method-not-found.
+  await waitForPapiMethodRegistered(`command:${commandName}`, WEBSOCKET_PORT, 60_000);
+  return sendPapiRequestOnce(`command:${commandName}`, args, WEBSOCKET_PORT, COMMAND_TIMEOUT_MS);
+}
+
+/**
+ * Make the installed sample WEB project editable. Its Settings.xml ships `<Editable>F</Editable>`,
+ * so `openScriptureEditor` would silently fall back to a read-only editor (main.ts overrides
+ * isReadOnly from `platform.isEditable`) — and a read-only Lexical editor never moves the caret on
+ * click, so click-follow cannot be exercised without this. Flipping the setting through the PDP
```
Comment (irahopkinson) [inline e2e-tests/tests/isolated/scroll-groups/scroll-group-sync.spec.ts:63]: This isn't quite true. Clicking in a readonly editor does move the selection, but the caret just isn't visible. So clicking in a readonly editor should update the scrRef.
